File: 프로그래머스/lv2/42578. 의상/의상.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

int solution(vector<vector<string>> clothes) {
        int answer = 1;
    //unordered_map<string, string> map;
    unordered_map<string, int> clothes_map; // 옷의 종류 - 갯수
    vector<string> clothes_type;
    for(vector<string> v : clothes) {
        //map.insert(make_pair(v[0],v[1])); // key : 옷의 이름 , value : 옷의 종류
        if(clothes_map.find(v[1]) != clothes_map.end()) {
            clothes_map[v[1]]++;
        }
        else {
            clothes_map.insert(make_pair(v[1],2)); //입는경우 + 이 종류의 옷을 아예 입지 않는경우
            clothes_type.push_back(v[1]);
        }
    }

    //경우의 수 따지는 방법
    //일단 맵에 있는 키의 개수 ( 각 하나씩이니까), 
    // 그리고 value의 개수에 따라 달라지는고임... 2개일 때는 ~ value개일때는
    //2개일때는 
    //아니면 종류마다 해시맵응ㄹ 만들어서... 
    int count = clothes_type.size();
    for(int i = 0; i<count; i++) {
        answer *= clothes_map[clothes_type[i]];
        
    }
    return answer-1;
}

File: 프로그래머스/lv2/42578. 의상/의상.js
function solution(clothes) {
    let answer = 1;
    //각 종류 별로 옷은 입거나, 안 입거나 둘 중 하나의 경우로 나뉜다.
    //입는 경우, 해당 종류의 옷 개수로 경우의 수가 생긴다.
    //ex : A종류 옷 1벌 - 입기/안입기 총 2가지
    //A종류 옷 2벌 - 입기(a1입기/a2입기)/안입기 총 3가지
    //A종류 옷 3벌 - 입기(a1입기/a2입기/a3입기)/안입기 총 4가지
    // 각 종류별 (옷의 개수 + 1) 끼리 전부 곱하면 경우의 수 모두 완성...
    
    const map = new Map();
    clothes.map(cloth=>cloth[1]).forEach((type, index)=> {
        if(map.has(type)) {
            map.set(type,map.get(type)+1)
        }
        else {
            map.set(type,2);
        }
    })
    
    for(const value of map.values()) {
        answer *= value;
    }
    
    return answer-1; //최소 한 개의 의상은 입어야 하므로 아무 것도 안 입는 경우 제외
}